crates/integrations/datafusion/src/table.rs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use datafusion::datasource::TableProvider;
+use iceberg::{table::Table, Catalog, NamespaceIdent, Result, TableIdent};
+
+pub(crate) struct IcebergTableProvider {
+    _inner: Table,
+}
+
+impl IcebergTableProvider {
+    pub(crate) async fn try_new(
+        client: Arc<dyn Catalog>,
+        namespace: NamespaceIdent,
+        name: impl Into<String>,
+    ) -> Result<Self> {
+        let name = name.into();
+        let ident = TableIdent::new(namespace, name);
+        let table = client.load_table(&ident).await?;
+
+        Ok(IcebergTableProvider { _inner: table })
+    }
+}
+
+impl TableProvider for IcebergTableProvider {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> datafusion::arrow::datatypes::SchemaRef {
+        todo!()
+    }
+
+    fn table_type(&self) -> datafusion::datasource::TableType {
+        todo!()
+    }
+
+    fn scan<'life0, 'life1, 'life2, 'life3, 'async_trait>(
+        &'life0 self,
+        _state: &'life1 datafusion::execution::context::SessionState,
+        _projection: Option<&'life2 Vec<usize>>,
+        _filters: &'life3 [datafusion::prelude::Expr],
+        _limit: Option<usize>,
+    ) -> core::pin::Pin<
+        Box<
+            dyn core::future::Future<
+                    Output = datafusion::error::Result<
+                        Arc<dyn datafusion::physical_plan::ExecutionPlan>,
+                    >,
+                > + core::marker::Send
+                + 'async_trait,
+        >,
+    >
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        'life3: 'async_trait,
+        Self: 'async_trait,
+    {
+        todo!()
```

Comment:
I think this could be implemented with current scan api without filter pushdown?

Reply:
...I think this has to be done (PTAL @liurenjie1024):

- impl asyn fn scan() on IcebergTableProvider which has to return an `Arc<dyn Execution Plan>`
  - therefore we need to: impl the trait ExecutionPlan for IcebergTableScan which is basically a wrapper around an actual PhysicalPlan e.g. ParquetExec 
    - therefore we need to: create an PhysicalPlan when building our IcebergTableScan

I think creating the PhysicalPlan and providing all the input like FileScanConfig will be the majority of the work and I'm not sure we can provide all of this yet...but well see...

reference from delta-rs:
- [DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L790-L798)
- [ExecutionPlan for DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L814-L881)
- [building DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L507-L651) and the actual PhysicalPlan // [ParquetExec](https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/parquet/struct.ParquetExec.html)

Reply:
I took a look at the `ExecutionPlan` trait, and I think the main part if the `execute` method which could return an arrow record batch stream, which has been implemented by `TableScan::to_arrow`. Other parts, such as partitioning, ordering we can leave it as simple as possible. 
I agree that this is not blocker and we can move it to next pr.

Reply:
I don't think this will work?

```rust
fn execute(...) -> Result<SendableRecordBatchStream> {
  ...
  // I can't await here since fn execute() needs to remain sync
  // using tokio::Runtime block_on() seems weird?
  table_scan.to_arrow().await.unwrap()
  // also the return type is BoxStream<'static, Result<RecordBatch>>
  // which does not satisfy SendableRecordBatchStream since we need to
  // impl RecordBatchStream as well
}
```

...so, how to avoid those issues? 

We could try to impl RecordBatchStream on TableScan itself, this however would require us to have datafusion as a dependency in the core and we'd still have the async issue - so I don't think this would be a good solution. 

Or we try the same approach as [delta-rs](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L627-L642), and create an actual datafusion physical plan and wrap this:

```rust
struct IcebergTableScan {
  plan: Arc<dyn ExecutionPlan>
  ...
}

impl ExecutionPlan for IcebergTableScan {
  ...
  fn execute(...) -> Result<SendableRecordBatchStream> {
     self.plan.execute()  
  }
}
```

... what do you think @liurenjie1024 ?

Reply:
I think you can do what is documented here: https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#lazily-async-create-a-stream

So inside `IcebergTableProvider::scan` you create `IcebergScanExec` where `IcebergScanExec` implements `ExecutionPlan`.

The `ExecutionPlan::execute` impl of `IcebergScanExec` would be something like this

```rust
impl ExecutionPlan for IcebergScanExec {
  fn execute(
        &self,
        partition: usize,
        context: Arc<TaskContext>
    ) -> Result<SendableRecordBatchStream> {
    let fut = create_batch_stream(self.table_scan);

    let stream = futures::stream::once(fut).try_flatten();
    Ok(Box::pin(RecordBatchStreamAdapter::new(self.schema.clone(), stream)))
  }
}
```

/// async function that returns a stream
async fn create_batch_stream(table: TableScan) -> Result<SendableRecordBatchStream> {
    todo!()
}


(Not @liurenjie1024 but hopefully it is helpful - I didn't compile it :) )

Reply:
@simonvandel thanks a lot - this looks promising

Reply:
Thanks @simonvandel for the comments! Yeah, that looks great to me.